### ISSUE TITLE
swagger.json is always available at the root path.

### ISSUE THREAD

**Issue by vburenin** [go-swagger/go-swagger#477]
## Problem statement

swagger.json document is always available at the root path as http://1.1.1.1/swagger.json
I can see why, but I would be happier if it was prefixed by a basePath such as:

http://1.1.1.1/v1/swagger.json

**Comment by njsrikar** [comment]
hi,
Any chance you know the priority on this?
I am trying to get swagger.json be available on the same path as the api's basePath. Is it possible on current go-swagger implementation?

Thanks
Srikar

**Comment by gregmarr** [comment]
How very topical.  I'm just starting to look at this myself.

**Comment by gregmarr** [comment]
This looks promising:

```
// Spec creates a middleware to serve a swagger spec.
// This allows for altering the spec before starting the http listener.
// This can be useful if you want to serve the swagger spec from another path than /swagger.json
//
func Spec(basePath string, b []byte, next http.Handler) http.Handler {
```

**Comment by gregmarr** [comment]
Rats, doesn't look like that's actually used.

**Comment by gregmarr** [comment]
So this function is called by 

```
// APIHandler returns a handler to serve the API, this includes a swagger spec, router and the contract defined in the swagger spec
func (c *Context) APIHandler(builder Builder) http.Handler {
...
	return Spec("", c.spec.Raw(), Redoc(redocOpts, c.RoutesHandler(builder)))
}
```

and that is called by

```
func (o *<GENERATED>API) Serve(builder middleware.Builder) http.Handler {
	o.Init()

	if o.Middleware != nil {
		return o.Middleware(builder)
	}
	return o.context.APIHandler(builder)
}
```

So maybe the path could be added to the Context object.

**Comment by njsrikar** [comment]
thanks.
thats correct, but these files will get overwritten if/when you run the generator again.

**Comment by gregmarr** [comment]
Yes, I wasn't thinking for post-generate editing, just strategy for implementation.

**Comment by njsrikar** [comment]
thanks for the pointer though, I was able to get it working by using Spec in setupGlobalMiddleware()
`if r.URL.Path == swaggerJSONPath {
			swaggerSpec, err := loads.Analyzed(SwaggerJSON, "")
			if err != nil {
				logger.Log.Error("Failed to get swagger spec json.", nil)
			}
			rawSpec := swaggerSpec.Raw()
			rootHandler := middleware.Spec(path+"/ui/", rawSpec, http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
				rw.WriteHeader(http.StatusFound)
				return
			}))
			rootHandler.ServeHTTP(w, r)
			return
		}`

**Comment by casualjim** [comment]
you can also use --skip-main in the generate command so then the main file doesn't get overwritten
Also you can easily make your own main file. I generate one by default more as an example. Everything is exported so it's easy to make other schemes.
The swagger 2.0 convention is to make the spec for your domain available at the root so that it can be discovered by convention

**Comment by gregmarr** [comment]
I did like @njsrikar did and added the `Spec` call through `setupGlobalMiddleware` after saving the base path in main.

The convention doesn't work for me because my API is behind an API gateway, where there is a path assigned to my service, and the root of my service is not accessible.  For ease of using a single Swagger file to both provision the gateway and drive `go-swagger`, that base path is used in the swagger file and thus in my generated routes.

**Comment by njsrikar** [comment]
ditto with my case as well.

@casualjim thanks for the tip on --skip-main ..will come in handy.

**Comment by ccamel** [comment]
Same than @gregmarr, our REST API is behind an API gateway (Kong actually). As a workaround, 
I just needed to fix the location of the `swagger.json` in the html doc (the gateway strips the uri by default).

While waiting for a fix, it does the job.

```go
func setupGlobalMiddleware(handler http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path == "/docs" {
			redocOpts := middleware.RedocOpts{
				SpecURL: urlPrefix + "/swagger.json",
			}

			middleware.Redoc(redocOpts, handler).ServeHTTP(w, r)
			return
		}
		handler.ServeHTTP(w, r)
	})
}
```

**Comment by rdsubhas** [comment]
This is a huge source of pain for us. We run separate services under separate paths, like test.com/serviceA, test.com/serviceB, test.com/serviceC, etc. Serving the swagger json under root makes things really cumbersome...
